src/bin/sol-fbp-generator/main.c

```diff
@@ -757,6 +758,135 @@ collect_context_info(struct generate_context *ctx, struct fbp_data *data)
     return true;
 }
 
+static bool
+generate_memory_map_struct(const char *json_file, int *elements)
+{
+    struct sol_file_reader *fr = NULL;
+    struct sol_json_scanner scanner, entries;
+    struct sol_str_slice contents;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    uint32_t version = 0;
+
+    *elements = 0;
+
+    fr = sol_file_reader_open(json_file);
+    if (!fr) {
+        SOL_ERR("Couldn't open json file '%s': %s\n", json_file, sol_util_strerrora(errno));
+        return false;
+    }
+
+    contents = sol_file_reader_get_all(fr);
+    sol_json_scanner_init(&scanner, contents.data, contents.len);
+
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (!SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "maps"))
+            continue;
+
+        sol_json_scanner_init_from_token(&scanner, &value);
+        SOL_JSON_SCANNER_ARRAY_LOOP (&scanner, &token, SOL_JSON_TYPE_OBJECT_START, reason) {
+            struct sol_str_slice path = { };
+
+            out("static const struct sol_memmap_map _memmap%d = {\n", (*elements)++);
+
+            SOL_JSON_SCANNER_OBJECT_LOOP_NEST (&scanner, &token, &key, &value, reason) {
+
+                if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "path")) {
+                    sol_json_token_remove_quotes(&value);
+                    path = sol_json_token_to_slice(&value);
+                } else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "version")) {
+                    if (sol_json_token_get_uint32(&value, &version) < 0) {
+                        SOL_WRN("Couldn't get memory map version");
```

Comment:
it would be nice to print more context in the error messages, making it easier to debug and fix errors. Here I'd say we should keep the json offset in mind as well as providing some "path" in the JSON hierarchy (not to be confused with `path` key above -- however if there is a `path` slice, you can print it to add even more context).